### PR TITLE
Use context manager for ledger query

### PR DESCRIPTION
**Pull request by leafyoung**
For discussion

The ledger table has its associated columns / function, the use of context manager has a few advantages:
- shortens the code,
- creates a single entry .select() vs. various get_XXXX()
- creates "context" to implement joining further (not done yet).
- column name are of ColumnField type, with some tools to generate query string, like `.eq/.lt`, `.in_/.not_in/.between/.max_` and `.comma_()/.and_()/.or_()`. Automatic lower case matching for string value.

**Example**
```
with self.context.ledger.Block as bl:
    ledger_output = bl.select(
        columns=[bl.DIFFICULTY],
        limit=10,
        order_by=bl.NUMBER.desc()
```

* LegerQuery is a mix-in class combines query and table specification (e.g. column names).

### CODE REVIEW

**Comment by iwyrkore** [comment]
@leafyoung Generally I like the idea. I kind of expected to be using the Table subclass to run the queries but this is OK.

The `query_method` decorator is broken now. It used to be a called with the table name so it set the doc string but left the function calls unchanged by returning the function. Now it seems like it only sets the doc string if the function is actually called.

**Comment by newtectonics** [comment]
This has been sitting on the shelf for a month. Is it still relevant?

**Comment by leafyoung** [comment]
@iwyrkore, ready for review now. This updates both the ledger and contract ledger utility and added tests.

This commit has been tagged as 0.8.27.

Models have been updated and in https://github.com/credmark/credmark-models-py/pull/159

File: credmark/cmf/types/ledger_contract.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    Used by :class:`~credmark.cmf.types.ledger.ContractLedger` to query a
+    contract's function or event data.
+    You do not need to create an instance yourself.
+
+    Access an instance with ``contract.ledger.functions.contractFunctionName()``
+    or ``contract.ledger.events.contractEventName()`` where ``contractFunctionName``
+    and ``contractEventName`` are the actual names of functions and events
+    of the contract.
+
+    See the :meth:`~credmark.cmf.types.ledger.ContractLedger.ContractEntity.__call__`
```
Comment (iwyrkore) [inline credmark/cmf/types/ledger_contract.py:74]: This needs to be corrected for the `select` call... `~credmark.cmf.types.ledger_contract. ContractEntityQuery.select`

File: credmark/cmf/types/ledger_contract.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    """"""
+
+
+class ContractEntityQuery(LedgerQueryBase):
+    """
+    Used by :class:`~credmark.cmf.types.ledger.ContractLedger` to query a
+    contract's function or event data.
+    You do not need to create an instance yourself.
+
+    Access an instance with ``contract.ledger.functions.contractFunctionName()``
```
Comment (iwyrkore) [inline credmark/cmf/types/ledger_contract.py:69]: I guess this section is no longer correct.

**Pull request merged**